### PR TITLE
Assign status to processed archives

### PR DESCRIPTION
**Pull request by AyanSinhaMahapatra**
Reference: https://github.com/nexB/scancode.io/issues/827

### CODE REVIEW

File: scanpipe/pipes/d2d.py [hunk 1]
```diff
@@ -953,3 +953,32 @@ def _map_javascript_colocation_resource(
...
+    to_resources_archives = to_resources.no_status().filter(is_archive=True)
+
+    for to_archive in to_resources_archives:
+        archive_extract_path = to_archive.path + "-extract"
+        archive_extract_resource = to_resources.filter(path=archive_extract_path)
+
+        # TODO:There are archives which are not extracted by default, these need to
+        # be accounted for. See https://github.com/nexB/scancode.io/issues/827
+        if not archive_extract_resource:
+            continue
```
Comment (tdruez) [inline scanpipe/pipes/d2d.py:974]: Let's not leave TODOs in the code, those never get addressed. We want issues instead, but here it seems that we have one.

Comment (tdruez) [reply scanpipe/pipes/d2d.py:974]: Using `archive_extract_resource.exists()` would result in a more efficient query.

File: scanpipe/pipes/d2d.py [hunk 2]
```diff
@@ -953,3 +953,32 @@ def _map_javascript_colocation_resource(
...
+
+        archive_resources_unmapped = to_resources.no_status().filter(
+            path__startswith=archive_extract_path
+        )
+        # If there are resources in the archives which are unmapped,
+        # they are not considered as processed
+        if archive_resources_unmapped:
+            continue
```
Comment (tdruez) [inline scanpipe/pipes/d2d.py:982]: I think this is actually not needed and triggers an extra query, ie it evaluates the `archive_resources_unmapped` queryset with a SELECT query then does the UPDATE query.
Where doing directly the UPDATE (that may update nothing) would always ensure only 1 query is executed.
To be confirmed though.

**Review by tdruez — CHANGES_REQUESTED** [review]

**Comment by tdruez** [comment]
@AyanSinhaMahapatra Could you also add an entry in the changelog?

File: scanpipe/pipes/d2d.py (hunk 1 above)
Comment (AyanSinhaMahapatra) [reply scanpipe/pipes/d2d.py]: Thanks, removing the TODO there and using the exists() function.

File: scanpipe/pipes/d2d.py (hunk 2 above)
Comment (AyanSinhaMahapatra) [reply scanpipe/pipes/d2d.py]: I'm not sure I understand what you mean correctly.
The SELECT query and the UPDATE query are on two different set of files, one on `archive_resources_unmapped` and the other on a singular `to_archive`.
The SELECT query is done on all the files in the extracted package, to check if any of them still have no status.
Then if and only there are no files which have status empty in those extracted files, we do an UPDATE on only one resource, which is the resource for the archive in question.
But I can use the `.exists()` function here too as this is same as the case mentioned in https://github.com/nexB/scancode.io/pull/861#discussion_r1295798243 to make this more efficient?

**Comment by AyanSinhaMahapatra** [comment]
@tdruez thanks for the improvement suggesttions!
I've added the updates and a CHANGELOG entry, except this point: https://github.com/nexB/scancode.io/pull/861#discussion_r1295861240
Ready for your review again! :smile:

File: scanpipe/pipes/d2d.py (hunk 2 above)
Comment (tdruez) [reply scanpipe/pipes/d2d.py]: Nevermind, I've read too fast thinking the update was applying to the previous queryset.
All good ;)

**Pull request merged**